panic if a subcommand is wrapped in double quotes and contains spaces

This issue was originally reported as https://github.com/hashicorp/terraform/issues/16642, debugging for a while, it turns out to be a bug of `mitchellh/cli`.

Panic will occur under the following conditions:

- A `CLI.commandNested` is `true`
- The subcommand is wrapped in double quotes and contains spaces ( `./app "foo bar"` )
- If there is no double quote, it is a valid subcommand ( `./app foo bar` )

Minimum reproduction code as follows:

```main.go
package main

import (
	"log"
	"os"

	"github.com/mitchellh/cli"
)

type FooBar struct{}

func (f *FooBar) Help() string {
	return "app foo bar"
}

func (f *FooBar) Run(args []string) int {
	log.Println("Foo Bar!")
	return 0
}

func (f *FooBar) Synopsis() string {
	return "Print \"Foo Bar!\""
}

func main() {
	c := cli.NewCLI("app", "1.0.0")
	c.Args = os.Args[1:]
	c.Commands = map[string]cli.CommandFactory{
		"foo bar": func() (cli.Command, error) {
			return &FooBar{}, nil
		},
	}

	exitStatus, err := c.Run()
	if err != nil {
		log.Println(err)
	}

	os.Exit(exitStatus)
}
```

```
$ go run main.go foo bar
2017/11/26 22:13:50 Foo Bar!

$ go run main.go "foo bar"
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/mitchellh/cli.(*CLI).processArgs(0xc420090000)
        /Users/m-morita/src/github.com/mitchellh/cli/cli.go:667 +0xb8c
github.com/mitchellh/cli.(*CLI).init(0xc420090000)
        /Users/m-morita/src/github.com/mitchellh/cli/cli.go:382 +0x3b9
github.com/mitchellh/cli.(*CLI).(github.com/mitchellh/cli.init)-fm()
        /Users/m-morita/src/github.com/mitchellh/cli/cli.go:158 +0x2a
sync.(*Once).Do(0xc4200900b8, 0xc42004def8)
        /usr/local/Cellar/go/1.9.1/libexec/src/sync/once.go:44 +0xbe
github.com/mitchellh/cli.(*CLI).Run(0xc420090000, 0xc420074270, 0x11858dc, 0x7)
        /Users/m-morita/src/github.com/mitchellh/cli/cli.go:171 +0x6d
main.main()
        /Users/m-morita/work/tmp/20171124/main.go:34 +0x13f
```

Panic occurs in the following lines.

https://github.com/mitchellh/cli/blob/65fcae5817c8600da98ada9d7edf26dd1a84837b/cli.go#L667

If wrapping in double quotes then `len(c.Args) = 1` , but the args matches `c.commandTree.LongestPrefix (searchKey)`, so the index `i` is added by the number of spaces, resulting in out-of-range access.

I think this is an edge case, but it is not good to panic by user input.
How should we fix it?

--- a/cli.go
+++ b/cli.go
@@ -646,9 +646,29 @@ func (c *CLI) processArgs() {
 		if c.subcommand == "" && arg != "" && arg[0] != '-' {
 			c.subcommand = arg
 			if c.commandNested {
+				// If the command has a space in it, then it is invalid.
+				// Set a blank command so that it fails.
+				if strings.ContainsRune(arg, ' ') {
+					c.subcommand = ""
+					return
+				}
+
+				// Determine the argument we look to to end subcommands.
+				// We look at all arguments until one has a space. This
+				// disallows commands like: ./cli foo "bar baz". An argument
+				// with a space is always an argument.
+				j := len(c.Args) - 1
+				for k, v := range c.Args[i:] {
+					if strings.ContainsRune(v, ' ') {
+						break
+					}
+
+					j = i + k + 1
+				}
+
 				// Nested CLI, the subcommand is actually the entire
 				// arg list up to a flag that is still a valid subcommand.
-				searchKey := strings.Join(c.Args[i:], " ")
+				searchKey := strings.Join(c.Args[i:j], " ")
 				k, _, ok := c.commandTree.LongestPrefix(searchKey)
 				if ok {
 					// k could be a prefix that doesn't contain the full

--- a/cli_test.go
+++ b/cli_test.go
@@ -306,6 +306,94 @@ func TestCLIRun_nestedMissingParent(t *testing.T) {
 	}
 }
 
+func TestCLIRun_nestedNoArgs(t *testing.T) {
+	command := new(MockCommand)
+	cli := &CLI{
+		Args: []string{"foo", "bar"},
+		Commands: map[string]CommandFactory{
+			"foo": func() (Command, error) {
+				return new(MockCommand), nil
+			},
+			"foo bar": func() (Command, error) {
+				return command, nil
+			},
+		},
+	}
+
+	exitCode, err := cli.Run()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if exitCode != command.RunResult {
+		t.Fatalf("bad: %d", exitCode)
+	}
+
+	if !command.RunCalled {
+		t.Fatalf("run should be called")
+	}
+
+	if !reflect.DeepEqual(command.RunArgs, []string{}) {
+		t.Fatalf("bad args: %#v", command.RunArgs)
+	}
+}
+
+func TestCLIRun_nestedQuotedCommand(t *testing.T) {
+	command := new(MockCommand)
+	cli := &CLI{
+		Args: []string{"foo bar"},
+		Commands: map[string]CommandFactory{
+			"foo": func() (Command, error) {
+				return new(MockCommand), nil
+			},
+			"foo bar": func() (Command, error) {
+				return command, nil
+			},
+		},
+	}
+
+	exitCode, err := cli.Run()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if exitCode != 127 {
+		t.Fatalf("bad: %d", exitCode)
+	}
+}
+
+func TestCLIRun_nestedQuotedArg(t *testing.T) {
+	command := new(MockCommand)
+	cli := &CLI{
+		Args: []string{"foo", "bar baz"},
+		Commands: map[string]CommandFactory{
+			"foo": func() (Command, error) {
+				return command, nil
+			},
+			"foo bar": func() (Command, error) {
+				return new(MockCommand), nil
+			},
+		},
+	}
+
+	exitCode, err := cli.Run()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if exitCode != command.RunResult {
+		t.Fatalf("bad: %d", exitCode)
+	}
+
+	if !command.RunCalled {
+		t.Fatalf("run should be called")
+	}
+
+	if !reflect.DeepEqual(command.RunArgs, []string{"bar baz"}) {
+		t.Fatalf("bad args: %#v", command.RunArgs)
+	}
+}
+
 func TestCLIRun_printHelp(t *testing.T) {
 	testCases := [][]string{
 		{"-h"},

